### PR TITLE
GLv3-3202-frontend-structure

### PR DESCRIPTION
**Pull request by KH-Amir-TN**
Frontend:
	          - Set a new project structure for the front end JavaScript code influenced by the
	          react-router-demo https://stackblitz.com/edit/github-agqlf5/ :
	              -> main.jsx: An entry point where Routes, Layouts, Global context is provided.
	              -> app.jsx: The parent route component for the `/` endpoint.
	              -> components/: A directory where custom components resides.
	              -> components/styling: A directory where custom components with no business logic resides (i.e. custom buttons).
	              -> components/forms: A directory where custom forms and form related components resides.
	              -> routes: A directory where sub routes resides.
	          - Created a `comonents/styling/buttons/ButtonLink.jsx` component that wraps/connects a react bootstrap button component with the react router API.

                  TODO:
  		    - Create the signup form.
## Description

Defined a project structure for GLv3 frontend code.
## Testing Steps

1. Pull the code.
2. Install the dependencies `bundle install && npm|yarn install`.
3. Clean the previous assets build by running `rm app/assets/builds/*` (This won't remove .keep since it's hidden).
4. Clean the database and tmp files for a better isolation by running `rails tmp:clear && rails db:schema:cache:clear && rails db:migrate:reset`
5. Run the linter and specs `bundle exec rubocop --parallel && bundle exec rspec`
6. Run `./bin/dev` to run the assets builders processes and the Puma server all at once.

To test the API:
    1. Make a POST request to `/api/v1/users.json`.
     1.1 The request has to have an `Accept: application/json` header.
     1.2. The request body should have a JSON representation with this format: 
     `{ user: { :name, :email, :password, :password_confirmation } }` .
     -> When not specifying the `.json` format the API should render back the `components#index` so that CSR handles the request.
     -> When not specifying the `Accept` header correctly the same scenario should occur.
     -> For unknown routes in the server side the CSR should take charge of handling it.
     -> Not respecting the user data required format will return a 400 BAD request with a JSON object in the response body having an errors array with a message **Invalid or Missing parameters.**
     -> Any unpermitted values given in the user data will be filtered, only `name, email ,password,...` will pass.
 
 To test the UI:
    1.  Entering `/` should render this:    
image

    2. Entering `/signup` should render this:    
image

    3. Entering anything else should render this:
image

    4:  Clicking on the Signup button should have the same outcome as 2).

### CODE REVIEW

**Comment by KH-Amir-TN** [comment]
Pending for #3309 to be merged.

**Comment by hadicheaito1** [comment]
Should we make the .jsx files into .tsx files for consistency?

**Comment by hadicheaito1** [comment]
> Should we make the .jsx files into .tsx files for consistency?

Nvm, it seems that certain files need to be .jsx and not .tsx. Could someone explain why?

**Review by hadicheaito1 — DISMISSED** [review]

**Comment by KH-Amir-TN** [comment]
> > Should we make the .jsx files into .tsx files for consistency?
> 
> Nvm, it seems that certain files need to be .jsx and not .tsx. Could someone explain why?

My bad I haven't made sure that all files have the same extension.
We'll be using Typescript (TS) so the extensions should be `.tsx` indicating the use of TS along with JSX.
Since, I still haven't used TS yet all of the JavaScript (JS) files that I wrote are in plain JSX.
So that I stay correct and to further indicate the lack of TS I've explicated that with the `.jsx` extension.

I'll make sure to change all of JS files extension to `.jsx` and come back to them once I include TS.

File: app/javascript/app.tsx [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import React from "react";
```
Comment (KH-Amir-TN) [inline app/javascript/app.tsx:1]: Inconsistent Extension.

Comment (KH-Amir-TN) [reply app/javascript/app.tsx:1]: Done,

**Review by farhatahmad — APPROVED** [review]

**Pull request merged**